### PR TITLE
Improve database tests, try to make them clean up locks

### PR DESCRIPTION
**Pull request by Martin2112**
Try to clean up transactions in database tests. And in some of the more complex cases make the test fail if the tx is left open. Add a bit more structure to some of the bigger tests.

### CODE REVIEW

**Comment by Martin2112** [comment]
Motivation: We're going to need to get a list of logs to be sequenced (PR in progress). In tests this means putting the db in a clean state, which we can't do if locks are held.

File: storage/mysql/tree_storage_test.go [hunk 1]
```diff
@@ -242,115 +283,142 @@ func TestDequeueLeaves(t *testing.T) {
...
-	tx := beginTx(s, t)
-
-	leaves := createTestLeaves(leavesToInsert, 20)
 
-	if err := tx.QueueLeaves(leaves); err != nil {
-		t.Fatalf("Failed to queue leaves: %v", err)
-	}
+	{
+		tx := beginTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeLeaves", tx)
```
Comment (AlCutter) [inline storage/mysql/tree_storage_test.go]: Weird indentation here

File: storage/mysql/tree_storage_test.go [hunk 2]
```diff
@@ -242,115 +283,142 @@ func TestDequeueLeaves(t *testing.T) {
...
 	leavesToDequeue1 := 3
 	leavesToDequeue2 := 2
-	leaves := createTestLeaves(leavesToInsert, 20)
 
-	if err := tx.QueueLeaves(leaves); err != nil {
-		t.Fatalf("Failed to queue leaves: %v", err)
-	}
+	{
+		tx := beginTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTwoBatches", tx)
```
Comment (AlCutter) [inline storage/mysql/tree_storage_test.go]: indentation here too

File: storage/mysql/tree_storage_test.go [hunk 3]
```diff
@@ -242,115 +283,142 @@ func TestDequeueLeaves(t *testing.T) {
...
 
-	ensureAllLeafHashesDistinct(leaves3, t)
+		if err != nil {
+			t.Fatalf("Failed to dequeue leaves: %v", err)
+		}
 
-	// Plus the union of the leaf batches should all have distinct hashes
-	leaves2 = append(leaves2, leaves3...)
-	ensureAllLeafHashesDistinct(leaves2, t)
+		if len(leaves3) != leavesToDequeue2 {
```
Comment (AlCutter) [inline storage/mysql/tree_storage_test.go]: indentation

File: storage/mysql/tree_storage_test.go [hunk 4]
```diff
@@ -485,15 +557,18 @@ func TestLatestSignedLogRoot(t *testing.T) {
...
-	root2, err := tx.LatestSignedLogRoot()
+	{
+		tx2 := beginTx(s, t)
+		defer tx2.Rollback()
+		root2, err := tx2.LatestSignedLogRoot()
 
-	if err != nil {
-		t.Fatalf("Failed to read back new log root: %v", err)
-	}
+		if err != nil {
```
Comment (AlCutter) [inline storage/mysql/tree_storage_test.go]: indentation

File: storage/tree_storage.go [hunk 5]
```diff
@@ -19,6 +19,11 @@ type TreeTX interface {
 
 	// Rollback aborts any performed operations. No updates must be applied to the underlying storage.
 	Rollback() error
+
+	// Open indicates if this transaction is open. An open transaction is one for which
+	// Commit() or Rollback() has never been called. Implementations must do all clean up
+	// in these methods so transactions are assumed closed regardless of the reported success.
+	Open() bool
```
Comment (AlCutter) [inline storage/tree_storage.go]: Did you consider something like `IsOpen()`?  Could be confusing if it's taken as a verb.

File: storage/mysql/tree_storage_test.go (hunk 1 above)
Comment (Martin2112) [inline storage/mysql/tree_storage_test.go]: Strange. It looks OK in the editor and gofmt doesn't alter it. Same for other cases.

File: storage/tree_storage.go (hunk 5 above)
Comment (Martin2112) [inline storage/tree_storage.go]: I'd much prefer IsOpen but Get in Getters isn't idiomatic apparently. I'm going to make it IsOpen anyway.

**Pull request merged**